File: pkg/docstate/snapshot.go
package docstate

import (
	"bytes"
	"crypto/sha256"
	"database/sql"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"math"
	"time"

	"github.com/klauspost/compress/zstd"
	"rune/pkg/editor/buffer"
)

func (s *Store) PutBlob(content string) (string, error) {
	sum := sha256.Sum256([]byte(content))
	hash := hex.EncodeToString(sum[:])

	var buf bytes.Buffer
	enc, err := zstd.NewWriter(&buf)
	if err != nil {
		return "", fmt.Errorf("put blob: create zstd encoder: %w", err)
	}
	if _, err = enc.Write([]byte(content)); err != nil {
		return "", fmt.Errorf("put blob: compress: %w", err)
	}
	if err = enc.Close(); err != nil {
		return "", fmt.Errorf("put blob: finalize zstd: %w", err)
	}
	compressed := buf.Bytes()

	_, err = s.perm.Exec(
		`INSERT OR IGNORE INTO blobs(hash, content) VALUES(?,?)`,
		hash, compressed,
	)
	if err != nil {
		return "", fmt.Errorf("put blob %s: %w", hash, err)
	}
	return hash, nil
}

// GetBlob decompresses and returns the content stored under hash, verifying
// its SHA-256 against hash before returning (blob rot / bit-flip detection —
// closes the "GetBlob never re-verifies SHA-256" structural gap). A mismatch
// is a corrupt blob and is surfaced as an error, never silently returned.
func (s *Store) GetBlob(hash string) (string, error) {
	var compressed []byte
	err := s.perm.QueryRow(`SELECT content FROM blobs WHERE hash=?`, hash).Scan(&compressed)
	if err != nil {
		return "", fmt.Errorf("get blob %s: %w", hash, err)
	}

	dec, err := zstd.NewReader(bytes.NewReader(compressed))
	if err != nil {
		return "", fmt.Errorf("get blob %s: create zstd decoder: %w", hash, err)
	}
	defer dec.Close()

	data, err := io.ReadAll(dec)
	if err != nil {
		return "", fmt.Errorf("get blob %s: decompress: %w", hash, err)
	}

	sum := sha256.Sum256(data)
	if got := hex.EncodeToString(sum[:]); got != hash {
		return "", fmt.Errorf("get blob %s: content hash mismatch (corrupt blob): got %s", hash, got)
	}
	return string(data), nil
}

// CreateSnapshot stores a PURE recovery anchor for docID at the current
// journal position (seq), tagged with THIS Store's own session_id (v10) —
// a snapshot anchors ONE session's own replay window; two sessions editing
// the same docID keep entirely separate anchor chains. seq should be the
// most recently returned seq from AppendEdit so that RecoverDocument can
// find this snapshot as the closest anchor for any replay. No source
// taxonomy (Part III) — the disk fact and the 3-way-merge ancestor are
// served entirely by observations/saved_obs/ancestorAt; a snapshot's only
// job is bounding how far RecoverDocument ever has to replay.
func (s *Store) CreateSnapshot(docID int64, content string, seq int64) (int64, error) {
	hash, err := s.PutBlob(content)
	if err != nil {
		return 0, fmt.Errorf("create snapshot doc %d: %w", docID, err)
	}

	at := s.clock().UTC().Format(time.RFC3339Nano)
	res, err := s.perm.Exec(
		`INSERT INTO snapshots(doc_id, session_id, blob_hash, seq, created_at) VALUES(?,?,?,?,?)`,
		docID, s.sessionID, hash, seq, at,
	)
	if err != nil {
		return 0, fmt.Errorf("create snapshot doc %d: %w", docID, err)
	}

	id, err := res.LastInsertId()
	if err != nil {
		return 0, fmt.Errorf("create snapshot doc %d: last insert id: %w", docID, err)
	}
	return id, nil
}

// recoverAt reconstructs docID's content AS SEEN BY sessionID specifically
// (v10) — the session-scoped engine behind both RecoverDocument (always
// s.sessionID) and RecoverAcrossSessions/Load's cross-session read of an
// already-identified OTHER (confirmed-dead) session's content. Finds the
// most recent snapshot tagged with sessionID whose seq is ≤ that session's
// current undo position, then forward-replays only THAT session's own edit
// events between the snapshot seq and the current position using
// buffer.ReplayForward.
//
// Algorithm:
//  1. Read sessionID's own current_seq from session_documents (no row, or
//     NULL = at head → use MaxInt64).
//  2. Find newest snapshot tagged sessionID with seq ≤ targetSeq;
//     anchorContent = "" if none.
//  3. Gather sessionID's own edits from events with seq > anchorSnapshotSeq
//     AND seq ≤ targetSeq.
//  4. Apply buffer.ReplayForward(anchorContent, batches) and return.
func (s *Store) recoverAt(docID, sessionID int64) (string, error) {
	// Step 1: read sessionID's own current undo position.
	var nullableCS sql.NullInt64
	if err := s.perm.QueryRow(`SELECT current_seq FROM session_documents WHERE session_id=? AND doc_id=?`, sessionID, docID).Scan(&nullableCS); err != nil && err != sql.ErrNoRows {
		return "", fmt.Errorf("recover doc %d session %d: read current_seq: %w", docID, sessionID, err)
	}
	targetSeq := int64(math.MaxInt64)
	if nullableCS.Valid {
		targetSeq = nullableCS.Int64
	}

	// Step 2: find the nearest snapshot, TAGGED WITH sessionID, at or before
	// targetSeq. Break ties by id DESC (most-recently-written wins):
	// coalesced edits keep the SAME journal seq (AppendEdit updates the
	// existing event in place), so several snapshots can share one seq with
	// progressively newer content. A seq-only `ORDER BY seq DESC` picks an
	// arbitrary one at the tie and can anchor on STALE content, dropping the
	// latest coalesced keystroke from recovery (a §1.4.3 data-loss). id DESC
	// selects the freshest snapshot at that seq — the tie-break policy this
	// query alone implements (D7: the doc comment used to cite a
	// `LatestSnapshot` helper that no longer exists in this package).
	var anchorSnapshotSeq int64
	var anchorContent string
	var blobHash string
	err := s.perm.QueryRow(
		`SELECT seq, blob_hash FROM snapshots WHERE doc_id=? AND session_id=? AND seq <= ? ORDER BY seq DESC, id DESC LIMIT 1`,
		docID, sessionID, targetSeq,
	).Scan(&anchorSnapshotSeq, &blobHash)
	if err != nil && err != sql.ErrNoRows {
		return "", fmt.Errorf("recover doc %d session %d: find anchor snapshot: %w", docID, sessionID, err)
	}
	if err == nil {
		anchorContent, err = s.GetBlob(blobHash)
		if err != nil {
			return "", fmt.Errorf("recover doc %d session %d: get anchor blob: %w", docID, sessionID, err)
		}
	}

	// Step 3: gather sessionID's OWN edit batches between the anchor and the
	// target position — never a different session's, even for the same doc.
	batches, err := s.readEditBatches(
		`SELECT edits FROM events WHERE doc_id=? AND session_id=? AND seq > ? AND seq <= ? ORDER BY seq ASC`,
		docID, sessionID, anchorSnapshotSeq, targetSeq,
	)
	if err != nil {
		return "", fmt.Errorf("recover doc %d session %d: %w", docID, sessionID, err)
	}

	// Step 4: replay edits from the anchor snapshot forward.
	return buffer.ReplayForward(anchorContent, batches), nil
}

// RecoverDocument reconstructs the current content for docID AS SEEN BY
// THIS SESSION (v10 — recoverAt, above). Two sessions editing the same
// docID each reconstruct only their own journal; neither can ever see the
// other's unsaved edits through this call.
func (s *Store) RecoverDocument(docID int64) (string, error) {
	return s.recoverAt(docID, s.sessionID)
}

// Content returns the current content of docID by reconstructing it from the
// VFS (snapshot + replayed edits). This is what autosave reads; disk is only
// written on an explicit materialise (⌘S).
func (s *Store) Content(docID int64) (string, error) {
	return s.RecoverDocument(docID)
}

// EditRow pairs one journaled edit batch with the seq it was recorded at.
type EditRow struct {
	Seq   int64
	Edits []buffer.AppliedEdit
}

// EditsInRange returns docID's own edit rows with seq in (fromSeq, toSeq],
// each tagged with its seq, ordered ascending — the same anchored-window
// shape recoverAt uses for a snapshot anchor (step 3), generalized to a
// caller-supplied bound instead of a snapshot's seq, and exposing per-row
// seq so a caller can identify the CURRENT TAIL row (the one at seq ==
// toSeq) — the only row AppendEdit's keystroke-coalescing UPDATE can still
// mutate in place (coalescing only ever targets the doc's current max-seq
// row) — from every row strictly before it, which stays valid to cache as
// long as the caller re-verifies CurrentSeq every call and evicts on any
// decrease: a strictly-older row can still be deleted later by undo-then-
// edit truncation, but that always requires an earlier, separate settle
// that lowers CurrentSeq below it first (undo and append never land in the
// same Update — see the fuzz driver's mirrorFor, the caller this exists
// for). Session-scoped (v10): each fuzzed Store/workspace pair is its own
// session, so this only ever sees its own edits, never a different
// session's sharing the same docID.
func (s *Store) EditsInRange(docID, fromSeq, toSeq int64) ([]EditRow, error) {
	rows, err := s.perm.Query(
		`SELECT seq, edits FROM events WHERE doc_id=? AND session_id=? AND seq > ? AND seq <= ? ORDER BY seq ASC`,
		docID, s.sessionID, fromSeq, toSeq,
	)
	if err != nil {
		return nil, fmt.Errorf("edits in range doc %d: %w", docID, err)
	}
	defer rows.Close()

	var result []EditRow
	for rows.Next() {
		var seq int64
		var editsJSON string
		if err := rows.Scan(&seq, &editsJSON); err != nil {
			return nil, fmt.Errorf("edits in range doc %d: scan: %w", docID, err)
		}
		var batch []buffer.AppliedEdit
		if err := json.Unmarshal([]byte(editsJSON), &batch); err != nil {
			return nil, fmt.Errorf("edits in range doc %d: unmarshal: %w", docID, err)
		}
		result = append(result, EditRow{Seq: seq, Edits: batch})
	}
	if err := rows.Err(); err != nil {
		return nil, fmt.Errorf("edits in range doc %d: rows: %w", docID, err)
	}
	return result, nil
}

// HasHistory reports whether docID has any events or snapshots RECORDED BY
// THIS SESSION (v10). Use this to distinguish "this session has no VFS
// record yet" (false → fall back to disk, or to a DIFFERENT session's
// history via RecoverAcrossSessions/Load's cross-session inheritance) from
// "this session's VFS record exists" (true → use RecoverDocument even if
// content is empty, e.g. the user deleted all text and the deletion was
// journaled). A docID with lots of history under OTHER sessions still
// reports false here for a session that has never itself touched it.
func (s *Store) HasHistory(docID int64) (bool, error) {
	var n int
	err := s.perm.QueryRow(
		`SELECT EXISTS(
			SELECT 1 FROM events WHERE doc_id=? AND session_id=?
			UNION ALL
			SELECT 1 FROM snapshots WHERE doc_id=? AND session_id=?
		)`,
		docID, s.sessionID, docID, s.sessionID,
	).Scan(&n)
	if err != nil {
		return false, fmt.Errorf("has history doc %d: %w", docID, err)
	}
	return n > 0, nil
}

File: rust/crates/rune-db/src/snapshot.rs
//! Recovery-anchor snapshots and `recover_document`'s forward replay — port
//! of `pkg/docstate/snapshot.go:74-180`. A snapshot is a PURE recovery
//! anchor: it exists only to bound how far `recover_document` ever has to
//! replay, never a source-of-truth taxonomy (that lives in `observations`,
//! WP4). CONSTITUTION §1.4.10: every snapshot's content flows through
//! `blob::put_blob` before it's ever discarded from the journal's live
//! window (`journal::append_edit`'s future-truncation deletes both `events`
//! and `snapshots` rows, but the blob a surviving snapshot still points to
//! is untouched).

use std::time::SystemTime;

use rusqlite::{Connection, OptionalExtension, Transaction, params};

use rune_core::buffer::Buffer;
use rune_core::undo::reapply;

use crate::Error;
use crate::blob::{get_blob, put_blob};
use crate::journal::edits_in_range;
use crate::session::format_rfc3339_nanos;

/// Stores a recovery anchor for `doc_id` at journal position `seq`, tagged
/// with `session_id` — a snapshot anchors ONE session's own replay window;
/// two sessions editing the same `doc_id` keep entirely separate anchor
/// chains. `seq` should be the most recently returned seq from
/// `journal::append_edit` so `recover_document` can find this snapshot as
/// the closest anchor for any replay. Port of `snapshot.go:74-103`.
pub fn create_snapshot(
    tx: &Transaction<'_>,
    session_id: i64,
    now: SystemTime,
    doc_id: i64,
    content: &str,
    seq: i64,
) -> Result<i64, Error> {
    let hash = put_blob(tx, content.as_bytes())?;
    let at = format_rfc3339_nanos(now);
    tx.execute(
        "INSERT INTO snapshots(doc_id, session_id, blob_hash, seq, created_at) \
         VALUES(?1, ?2, ?3, ?4, ?5)",
        params![doc_id, session_id, hash, seq, at],
    )?;
    Ok(tx.last_insert_rowid())
}

/// Reconstructs `doc_id`'s content AS SEEN BY `session_id`:
///
/// 1. Read `session_id`'s own current undo position from
///    `session_documents` (no row, or NULL, means "at head" → use
///    `i64::MAX`).
/// 2. Find the newest snapshot tagged `session_id` with `seq <= target`,
///    tie-broken `seq DESC, id DESC` (coalesced edits keep the SAME seq, so
///    several snapshots can share one seq with progressively newer content
///    — `id DESC` picks the freshest one at that seq, never an arbitrary
///    tie); `anchor_content = ""` if none.
/// 3. Gather `session_id`'s own edit batches with `seq` in
///    `(anchor_seq, target]`.
/// 4. Forward-replay those batches onto `anchor_content`, one row at a
///    time, using `rune_core::undo::reapply`'s edit-application semantics
///    (ascending by `start` within a row, against a running buffer) — NOT
///    Go's `buffer.ReplayForward`, which silently clamps/skips an
///    out-of-range edit instead of erroring (`replay.go:26-33`). Every row
///    here was produced by this crate's own `append_edit`, so a replay
///    failure means a genuinely corrupt/inconsistent journal, which §1.3
///    requires surfacing rather than silently reconstructing wrong content.
///
/// Port of `snapshot.go:105-172` (`recoverAt`/`RecoverDocument`). Takes
/// `&Connection` rather than `&Transaction` — this is a pure read, so
/// `rusqlite`'s `Transaction: Deref<Target=Connection>` lets writer-thread
/// callers (inside `retry::with_retry`'s `&Transaction`) and read-only
/// callers (`reader.rs`, a plain `&Connection`) share this ONE
/// implementation instead of two copies of the same query sequence.
pub fn recover_document(conn: &Connection, session_id: i64, doc_id: i64) -> Result<String, Error> {
    let current_seq: Option<i64> = conn
        .query_row(
            "SELECT current_seq FROM session_documents WHERE session_id=?1 AND doc_id=?2",
            params![session_id, doc_id],
            |r| r.get::<_, Option<i64>>(0),
        )
        .optional()?
        .flatten();
    let target = current_seq.unwrap_or(i64::MAX);

    let anchor: Option<(i64, String)> = conn
        .query_row(
            "SELECT seq, blob_hash FROM snapshots \
             WHERE doc_id=?1 AND session_id=?2 AND seq <= ?3 \
             ORDER BY seq DESC, id DESC LIMIT 1",
            params![doc_id, session_id, target],
            |r| Ok((r.get(0)?, r.get(1)?)),
        )
        .optional()?;

    let (anchor_seq, anchor_content) = match anchor {
        Some((seq, hash)) => {
            let bytes = get_blob(conn, &hash)?;
            // The anchor blob is always session-authored buffer content
            // (written by `create_snapshot` from a `&str`) re-entering the
            // String-typed edit buffer here — a decode failure means a
            // genuinely corrupt snapshot, which must surface as an error,
            // never be silently coerced (blob.rs module doc).
            let content = String::from_utf8(bytes).map_err(|e| {
                Error::CorruptPayload(format!(
                    "snapshot blob {hash} for doc {doc_id}: non-utf8 content: {e}"
                ))
            })?;
            (seq, content)
        }
        None => (0, String::new()),
    };

    let rows = edits_in_range(conn, session_id, doc_id, anchor_seq, target)?;

    let mut buf = Buffer::new(anchor_content);
    for row in rows {
        buf = reapply(&buf, &row.edits).map_err(|e| {
            Error::ReplayFailed(format!(
                "doc {doc_id} session {session_id} at seq {}: {e}",
                row.seq
            ))
        })?;
    }
    Ok(buf.content().to_string())
}

#[cfg(test)]
#[allow(clippy::unwrap_used, clippy::expect_used, clippy::indexing_slicing)]
mod tests {
    use super::*;
    use crate::journal::{append_edit, move_undo_pos, undo_peek};
    use rune_core::buffer::AppliedEdit;

    fn open() -> Connection {
        let conn = Connection::open_in_memory().expect("open");
        crate::schema::apply(&conn).expect("schema");
        conn.pragma_update(None, "foreign_keys", "ON").expect("fk");
        conn
    }

    fn insert_test_document(tx: &Transaction<'_>) -> i64 {
        tx.execute(
            "INSERT INTO documents(path, created_at, last_seen_at) VALUES ('', 'x', 'x')",
            [],
        )
        .expect("insert document");
        tx.last_insert_rowid()
    }

    fn insert_test_session(conn: &Connection) -> i64 {
        crate::session::establish_session(conn, SystemTime::now()).expect("establish session")
    }

    fn text_insert(s: &str) -> Vec<AppliedEdit> {
        vec![AppliedEdit {
            start: 0,
            end: 0,
            deleted: String::new(),
            insert: s.to_string(),
        }]
    }

    #[test]
    fn recover_with_no_snapshot_replays_from_empty() {
        let mut conn = open();
        let session_id = insert_test_session(&conn);
        let tx = conn.transaction().expect("tx");
        let doc_id = insert_test_document(&tx);

        append_edit(
            &tx,
            session_id,
            SystemTime::now(),
            doc_id,
            &text_insert("hello"),
            &[],
            &[],
        )
        .expect("append_edit");

        let got = recover_document(&tx, session_id, doc_id).expect("recover");
        assert_eq!(got, "hello");
        tx.commit().expect("commit");
    }

    #[test]
    fn recover_uses_newest_anchor_at_or_before_target_and_replays_the_rest() {
        let mut conn = open();
        let session_id = insert_test_session(&conn);
        let tx = conn.transaction().expect("tx");
        let doc_id = insert_test_document(&tx);

        append_edit(
            &tx,
            session_id,
            SystemTime::now(),
            doc_id,
            &text_insert("file-edit-1"),
            &[],
            &[],
        )
        .expect("append_edit 1");
        let seq1 = crate::journal::current_seq(&tx, session_id, doc_id).expect("seq1");
        create_snapshot(
            &tx,
            session_id,
            SystemTime::now(),
            doc_id,
            "file-edit-1",
            seq1,
        )
        .expect("snapshot");

        append_edit(
            &tx,
            session_id,
            SystemTime::now(),
            doc_id,
            &text_insert("-more"),
            &[],
            &[],
        )
        .expect("append_edit 2");

        let got = recover_document(&tx, session_id, doc_id).expect("recover");
        assert_eq!(got, "-morefile-edit-1");
        tx.commit().expect("commit");
    }

    #[test]
    fn recover_respects_current_seq_after_undo() {
        let mut conn = open();
        let session_id = insert_test_session(&conn);
        let tx = conn.transaction().expect("tx");
        let doc_id = insert_test_document(&tx);

        append_edit(
            &tx,
            session_id,
            SystemTime::now(),
            doc_id,
            &text_insert("a"),
            &[],
            &[],
        )
        .expect("append a");
        append_edit(
            &tx,
            session_id,
            SystemTime::now(),
            doc_id,
            &text_insert("b"),
            &[],
            &[],
        )
        .expect("append b");

        let step = undo_peek(&tx, session_id, doc_id)
            .expect("undo_peek")
            .expect("something to undo");
        move_undo_pos(&tx, session_id, doc_id, step.new_pos).expect("move_undo_pos");

        let got = recover_document(&tx, session_id, doc_id).expect("recover");
        assert_eq!(got, "a", "recovery must stop at the undone position");
        tx.commit().expect("commit");
    }
}
